src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1723,19 +1723,35 @@ private void WaitForAllLoggingServiceEventsToBeProcessed()
         /// </summary>
         private void HandleSubmissionException(GraphBuildSubmission submission, Exception ex)
         {
-            if (ex is InvalidProjectFileException projectException)
+            if (ex is AggregateException ae)
             {
-                if (!projectException.HasBeenLogged)
+                // If there's exactly 1, just flatten it
+                if (ae.InnerExceptions.Count == 1)
                 {
-                    BuildEventContext buildEventContext = new BuildEventContext(submission.SubmissionId, 1, BuildEventContext.InvalidProjectInstanceId, BuildEventContext.InvalidProjectContextId, BuildEventContext.InvalidTargetId, BuildEventContext.InvalidTaskId);
-                    ((IBuildComponentHost)this).LoggingService.LogInvalidProjectFileError(buildEventContext, projectException);
-                    projectException.HasBeenLogged = true;
+                    ex = ae.InnerExceptions[0];
+                }
+                else
+                {
+                    // Log each InvalidProjectFileException encountered during ProjectGraph creation
+                    foreach (Exception innerException in ae.InnerExceptions)
+                    {
+                        if (innerException is InvalidProjectFileException innerProjectException)
+                        {
+                            LogInvalidProjectFileError(innerProjectException);
+                        }
+                    }
                 }
             }
 
-            ex = ex is AggregateException ae && ae.InnerExceptions.Count == 1
-                ? ae.InnerExceptions.First()
-                : ex;
+            if (ex is InvalidProjectFileException projectException)
+            {
+                LogInvalidProjectFileError(projectException);
+            }
+
+            if (ex is CircularDependencyException)
+            {
+                LogInvalidProjectFileError(new InvalidProjectFileException(ex.Message, ex));
```

Comment:
After the latest iteration, it looks like this:

```
MSBuild version 17.4.0-dev-22365-01+1223c9a6b for .NET Framework
Build started 7/15/2022 1:28:30 PM.
MSBUILD : error : MSB4251: There is a circular dependency involving the following projects:
MSBUILD : error : C:\Users\David\Code\tmp\a.proj ->
MSBUILD : error : C:\Users\David\Code\tmp\b.proj ->
MSBUILD : error : C:\Users\David\Code\tmp\a.proj

Build FAILED.

  MSBUILD : error : MSB4251: There is a circular dependency involving the following projects:
MSBUILD : error : C:\Users\David\Code\tmp\a.proj ->
MSBUILD : error : C:\Users\David\Code\tmp\b.proj ->
MSBUILD : error : C:\Users\David\Code\tmp\a.proj

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.25
```

I don't think there are line numbers to give really since this isn't like non-graph builds where the circular dependency comes from a specific `<MSBuild>` task call. I guess in a graph build it comes from a `<ProjectReference>`, but we don't really keep the xml elements around to figure out where those are.

Reply:
Compare with today's pretty awful behavior:

```
Microsoft (R) Build Engine version 17.2.1+52cd2da31 for .NET Framework
Copyright (C) Microsoft Corporation. All rights reserved.

Build started 7/15/2022 1:43:31 PM.

Build FAILED.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:00.03
MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
Microsoft.Build.Exceptions.CircularDependencyException: MSB4251: There is a circular dependency involving the following projects:
C:\Users\David\Code\tmp\a.proj ->
C:\Users\David\Code\tmp\b.proj ->
C:\Users\David\Code\tmp\a.proj
   at Microsoft.Build.Graph.GraphBuilder.<DetectCycles>g__VisitNode|32_0(ProjectGraphNode node, IDictionary`2 nodeState)
   at Microsoft.Build.Graph.GraphBuilder.DetectCycles(IReadOnlyCollection`1 entryPointNodes, ProjectInterpretation projectInterpretation, Dictionary`2 allParsedProjects)
   at Microsoft.Build.Graph.GraphBuilder.BuildGraph()
   at Microsoft.Build.Graph.ProjectGraph..ctor(IEnumerable`1 entryPoints, ProjectCollection projectCollection, ProjectInstanceFactoryFunc projectInstanceFactory, Int32 degreeOfParallelism, CancellationToken cancellationToken)
   at Microsoft.Build.Execution.BuildManager.ExecuteGraphBuildScheduler(GraphBuildSubmission submission)
   at Microsoft.Build.Execution.BuildManager.<>c__DisplayClass85_0.<ExecuteSubmission>b__0()

MSBUILD : error MSB1025: An internal failure occurred while running MSBuild.
Microsoft.Build.Exceptions.CircularDependencyException: MSB4251: There is a circular dependency involving the following projects:
C:\Users\David\Code\tmp\a.proj ->
C:\Users\David\Code\tmp\b.proj ->
C:\Users\David\Code\tmp\a.proj
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Boolean needToValidateProject, String schemaFile, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsNotAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache, String commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String commandLine)

Unhandled Exception: Microsoft.Build.Exceptions.CircularDependencyException: MSB4251: There is a circular dependency involving the following projects:
C:\Users\David\Code\tmp\a.proj ->
C:\Users\David\Code\tmp\b.proj ->
C:\Users\David\Code\tmp\a.proj
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Boolean needToValidateProject, String schemaFile, Int32 cpuCount, Boolean enableNodeReuse, TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsNotAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, Boolean isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, String[] inputResultsCaches, String outputResultsCache, String commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String commandLine)
   at Microsoft.Build.CommandLine.MSBuildApp.Main()
```